### PR TITLE
ARROW-9742: [Rust] [DataFusion] Improved DataFrame trait (formerly known as the Table trait)

### PR DESCRIPTION
**Pull request by andygrove**
This PR renames the `Table` trait to `DataFrame` and also contains the following improvements:

- Updates the `collect` method signature to remove the reference to an execution context (which made no sense because the table/dataframe struct was created by the context).
- Adds rustdocs for DataFrame and ExecutionContext with code examples
- Updates existing examples to call `ExecutionContext.sql()` instead of referencing logical and physical query plans - these are internal implementation details that should be of no concern to typical users
- Adds a new example that uses the DataFrame trait to build a query
- A few other minor UX improvements so that I could show some real examples

Note that the goal of this PR is mostly to document the API as it is today. There are number of design flaws that still need fixing and I will be creating follow-up PRs for those.

### CODE REVIEW

**Comment by andygrove** [comment]
@alamb @jorgecarleitao

**Comment by jorgecarleitao** [comment]
I went through this, and it looks great. I admit I am still not fully familiar when to use Rc vs Arc yet (e.g. why nor Arc instead? because we do not expect the context to be sharable between threads?), so take that into account.

I have an extra suggestion, that I pushed [here](https://github.com/jorgecarleitao/arrow/commit/521dd2be7644bb8b7bf5cb47164e1246bc45d13f). Essentially, with this change I think that we can remove `ExecutionContextSchemaProvider` altogether and let `ExecutionContextState` implement `SchemaProvider`.

File: rust/datafusion/examples/memory_table_api.rs [hunk 1]
```diff
@@ -51,15 +51,15 @@ fn main() -> Result<()> {
     // declare a table in memory. In spark API, this corresponds to createDataFrame(...).
     let provider = MemTable::new(schema, vec![vec![batch]])?;
     ctx.register_table("t", Box::new(provider));
-    let t = ctx.table("t")?;
+    let df = ctx.table("t")?;
```
Comment (jorgecarleitao) [inline rust/datafusion/examples/memory_table_api.rs:54]: Shouldn't we also rename `.table` and `.register_table`?

Comment (alamb) [reply rust/datafusion/examples/memory_table_api.rs:54]: FWIW I think `table` is still reasonable from a DataFusion point of view as the data is creating a logical table in the context of a query, while being backed by a Dataframe. Changing to `register_dataframe` would be fine too in my opinion. I could go either way

File: rust/datafusion/src/dataframe.rs [hunk 2]
```diff
@@ -15,42 +15,46 @@
...
 use crate::arrow::record_batch::RecordBatch;
 use crate::error::Result;
-use crate::execution::context::ExecutionContext;
 use crate::logicalplan::{Expr, LogicalPlan};
 use arrow::datatypes::Schema;
 use std::sync::Arc;
 
-/// Table is an abstraction of a logical query plan
-pub trait Table {
+/// DataFrame is an abstraction of a logical query plan
```
Comment (alamb) [inline rust/datafusion/src/dataframe.rs:26]:
```suggestion
/// DataFrame represents a logical set of rows with the same named columns.
/// Similar to a [Pandas DataFrame](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) or [Spark DataFrame](https://spark.apache.org/docs/latest/sql-programming-guide.html)
```

File: rust/datafusion/src/dataframe.rs [hunk 3]
```diff
@@ -15,42 +15,46 @@
...
 
     /// Return the logical plan
     fn to_logical_plan(&self) -> LogicalPlan;
 
-    /// Return an expression representing a column within this table
-    fn col(&self, name: &str) -> Result<Expr>;
+    /// Collects the result as a vector of RecordBatch.
+    fn collect(&self, batch_size: usize) -> Result<Vec<RecordBatch>>;
+
+    /// Returns the schema
```
Comment (alamb) [inline rust/datafusion/src/dataframe.rs:53]:
```suggestion
    /// Returns the schema (names and types of columns) in this DataFrame
```

File: rust/datafusion/src/execution/context.rs [hunk 4]
```diff
@@ -17,9 +17,11 @@
 
 //! ExecutionContext contains methods for registering data sources and executing queries
 
+use std::cell::RefCell;
 use std::collections::{HashMap, HashSet};
 use std::fs;
 use std::path::Path;
+use std::rc::Rc;
```
Comment (alamb) [inline rust/datafusion/src/execution/context.rs:24]: As mentioned by @jorgecarleitao  -- I suggest using `Arc` in this file unless there is a good reason to use `Rc`

File: rust/datafusion/src/execution/context.rs [hunk 5]
```diff
@@ -89,12 +91,18 @@ impl ExecutionConfig {
 }
 
 /// Execution context for registering data sources and executing queries
-pub struct ExecutionContext {
-    datasources: HashMap<String, Box<dyn TableProvider + Send + Sync>>,
-    scalar_functions: HashMap<String, Box<ScalarFunction>>,
+#[derive(Clone)]
+pub struct ExecutionContextState {
+    datasources: Rc<RefCell<HashMap<String, Box<dyn TableProvider + Send + Sync>>>>,
```
Comment (alamb) [inline rust/datafusion/src/execution/context.rs:96]: Can you explain the rationale for using `RefCell` here? It seems to me like `Arc<HashMap<...>>` is easier to reason about. I can't think of an example when running multi-threaded when you would want one thread to borrow mutably -- it seems like an ExecutionContextState would be created/configured by one thread, and then when shared would be read only.

**Review by alamb — COMMENTED** [review]
The changes in the PR description (cleanup / rename of `Table` --> `DataFrame`) make a lot of sense to me and are 👍 

>  (e.g. why nor Arc instead? because we do not expect the context to be sharable between threads?), 

I agree with @jorgecarleitao that we should use `Arc` unless we have some good reason not to -- doing so adds minimal additional overhead and allows sharing between threads and use in `async` code. 

Also, I don't understand the change to use `RefCell` -- normally I have seen that used when you have a struct that is immutable from its users perspective but need some mutable state under the covers (e.g. the ref count on a ref counted pointer).

File: rust/datafusion/src/execution/context.rs (hunk 5 above)
Comment (andygrove) [reply rust/datafusion/src/execution/context.rs:96]: So we do need interior mutability here. Users can register new data sources. I can convert to `Arc<Mutex<_>>` instead though.

File: rust/datafusion/examples/memory_table_api.rs (hunk 1 above)
Comment (andygrove) [reply rust/datafusion/examples/memory_table_api.rs:54]: SparkSession has a `table` method described as `Returns the specified table/view as a DataFrame`. I tend to follow Spark conventions where possible.

**Comment by andygrove** [comment]
@jorgecarleitao @alamb @houqp @jhorstmann This is ready for re-review. Please see the updated PR description.

File: rust/datafusion/src/dataframe.rs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    ///
+    /// let mut ctx = ExecutionContext::new();
+    /// let df = ctx.read_csv("tests/example.csv", CsvReadOptions::new()).unwrap();
+    /// let batches = df.collect(4096).unwrap();
+    /// ```
+    fn collect(&self, batch_size: usize) -> Result<Vec<RecordBatch>>;
+
+    /// Returns the schema describing the output of this DataFrame in terms of columns returned,
+    /// where each column has a name, data type, and nullability attribute.
+
```
Comment (jorgecarleitao) [inline rust/datafusion/src/dataframe.rs:139]: I am not sure if this won't break descriptions (add `    ///`?). If not, ignore this comment.

**Review by jorgecarleitao — APPROVED** [review]
LGTM! 

Really great work! I particularly liked that:

* all operations now expect `Expr` instead of `&Expr`: it avoids writing extra `&` on almost every statement
* `df.col` is no longer valid
* proper documentation and examples!

**Comment by andygrove** [comment]
Thanks @jorgecarleitao  I wanted to move the aggregate functions too but ran into a problem and it probably relates to your PRs .. I will file a JIRA to explain this.

**Comment by jorgecarleitao** [comment]
> Thanks @jorgecarleitao I wanted to move the aggregate functions too but ran into a problem and it probably relates to your PRs .. I will file a JIRA to explain this.

Yes, currently aggregate functions are impossible to correctly declare in the logical realm because they have a variable return type, while all logical expressions have a hard-coded return type. I propose a fix for that in one of my PRs. The planets are aligning!

File: rust/datafusion/src/execution/context.rs (hunk 5 above)
Comment (alamb) [reply rust/datafusion/src/execution/context.rs]: I guess I was thinking that users would register new datasources before/during query planning time (when `ExecutionContextState` was owned by a single thread and therefore the entire `struct` would be mutable, not just a field. I probably misunderstand what is going on

**Pull request closed without merging**